### PR TITLE
[dreamscreen][WIP] Initial contribution

### PR DESCRIPTION
**Pull request by markus7017**
The binding controls the [DreamScreen devices](https://www.dreamscreentv.com). DreamScreen provides Amibilight feature to TVs without that feature. Somehow comparable to Hyperion, but as ready-to-use kit with WiFi and an app to control the devices.

This binding integrates DreamScreen devices into openHAB.

Initial checkin for DreamScreen binding on behalf of Bruce Brouwer, the initial contributor of this binding.

Signed-off-by: Markus Michels <markus7017@gmail.com>

### CODE REVIEW

**Comment by J-N-K** [comment]
Thanks for your contribution. Please add documentation. Thanks. Marking at as work-in-progress until then. Remove that after you're done.

**Comment by D3ltorohd** [comment]
Hello, i have a dreamscreen 4k controller to. So when i can help testing that binding, i will do that.

**Comment by Hilbrand** [comment]
@markus7017 What is the status of this binding? The code looks fine. I can review if you want to finish this. Briefly looking a see only minor issues to improve. Most work would be to complete the README. 
You'll need to do a rebase (and push force) to fix the merge conflicts.

**Comment by markus7017** [comment]
@Hilbrand ok, let's go. I could update the README and do the rebase after finishing a small PR for Shelly.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    private synchronized void ensureRefreshing() throws IOException {
+        if (this.broadcastAddress == null) {
+            this.broadcastAddress = InetAddress.getByName("255.255.255.255");
+        }
+        if (this.refreshing == null) {
+            final DatagramSocket socket = new DatagramSocket(DREAMSCREEN_PORT, this.hostAddress);
+            socket.setReuseAddress(true);
+            this.serverSocket = socket;
+
+            final Thread server = new Thread(this::runServer, "dreamscreen-tv");
```
[the remark this answers is not in the document]
Comment (Hilbrand) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:101]: It don't see if this thread is every stopped?

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+                logger.error("Error receiving DreamScreen data", ioe);
+            }
+        }
+    }
+
+    void setHostAddress(@Nullable String value) {
+        try {
+            this.hostAddress = InetAddress.getByName(value);
+        } catch (UnknownHostException e) {
+            logger.error("Cannot set host address to {}", value, e);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:164]: If this happens this would probably mean the thing should be in offline status with this message?
```suggestion
            logger.debug("Cannot set host address to {}", value, e);
```

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    void send(int group, int commandUpper, int commandLower, byte[] payload, @Nullable InetAddress address)
+            throws IOException {
+        send(group, 0b00010001, commandUpper, commandLower, payload, address);
+    }
+
+    void send(int group, int flags, int commandUpper, int commandLower, byte[] payload, @Nullable InetAddress address)
+            throws IOException {
+        final DatagramSocket socket = new DatagramSocket();
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:224]: You can use try-with-resources and omit the close() call.
```suggestion
        try (final DatagramSocket socket = new DatagramSocket()) {
```

File: bundles/pom.xml [hunk 4]
```diff
@@ -1,4 +1,5 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (Hilbrand) [inline bundles/pom.xml:1]: I'm guessing the skeletonscript causes this change. Can you undo it.

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/i18n/dreamscreen_xx_XX.properties [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/i18n/dreamscreen_xx_XX.properties:1]: Remove this demo file

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenMessageConstants.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * The {@link DreamScreenMessageConstants} class defines common constants in relation to messages
+ *
+ * @author Bruce Brouwer - Initial contribution
+ */
+@NonNullByDefault
+public class DreamScreenMessageConstants {
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenMessageConstants.java:23]: This class is empty?

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandlerFactory.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Reference
+    protected void setNetworkAddressService(NetworkAddressService networkAddressService) {
+        server.setHostAddress(networkAddressService.getPrimaryIpv4HostAddress());
+    }
+
+    protected void unsetNetworkAddressService(NetworkAddressService networkAddressService) {
+        // nothing to really unset. This just needed to grab the primary IPv4 host address.
+    }
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandlerFactory.java:85]: Can you inject these via the constructor:
```
@Activate
public DreamScreenHandlerFactory(@Reference NetworkAddressService networkAddressService, @Reference DreamScreenDynamicStateDescriptionProvider provider) {
...
Then you can make the fields final too.
}

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        }
+    }
+
+    private void changePower(OnOffType command) {
+        try {
+            send(0x03, 0x01, new byte[] { command == ON ? powerOnMode.deviceMode : 0 });
+            this.powerOn = command == ON;
+            updateState(CHANNEL_POWER, command);
+        } catch (IOException e) {
+            logger.error("Error changing power state", e);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:129]:
```suggestion
            logger.debug("Error changing power state", e);
```

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    private void changeMode(DecimalType state) {
+        try {
+            final DreamScreenEnumMode mode = DreamScreenEnumMode.fromState(state);
+            if (this.powerOn) {
+                send(0x03, 0x01, new byte[] { mode.deviceMode });
+            }
+            this.powerOnMode = mode;
+            updateState(CHANNEL_MODE, state);
+        } catch (IOException e) {
+            logger.error("Error changing mode", e);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:144]:
```suggestion
            logger.debug("Error changing mode", e);
```

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
+        config = getConfigAs(DreamScreenConfiguration.class);
+        updateStatus(ThingStatus.UNKNOWN);
+        scheduler.execute(() -> {
+            boolean thingReachable = true; // <background task with long running initialization here>
+        try {
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:73]: Indentation. You can run `mvn spotless:apply` on your binding to fix any style issues.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    private void changeScene(DecimalType state) {
+        try {
+            final DreamScreenEnumScene scene = DreamScreenEnumScene.fromState(state);
+            send(0x03, 0x08, new byte[] { scene.deviceAmbientSceneType });
+            if (scene.deviceAmbientSceneType == 1) {
+                send(0x03, 0x0D, new byte[] { scene.deviceAmbientScene });
+            }
+            updateState(CHANNEL_SCENE, state);
+        } catch (IOException e) {
+            logger.error("Error changing scene", e);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:159]:
```suggestion
            logger.debug("Error changing scene", e);
```

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+                    "Cannot send scene command to " + getName());
+        }
+    }
+
+    private void changeInput(DecimalType state) {
+        try {
+            send(0x03, 0x20, new byte[] { state.byteValue() });
+            updateState(CHANNEL_INPUT, state);
+        } catch (IOException e) {
+            logger.error("Error changing input", e);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:170]:
```suggestion
            logger.debug("Error changing input", e);
```

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        } else {
+            logger.warn("DreamScreen {} is not on-line", getName());
+        }
+    }
+
+    void refreshState(final byte[] data, int off, int len, InetAddress address) {
+        this.address = address;
+        this.group = data[off + 38];
+
+        logger.info("{}", data.toString());
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:188]:
```suggestion
        logger.debug("{}", data.toString());
```
Or trace level.
Also add some context to this data.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+@NonNullByDefault
+public class DreamScreenHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(DreamScreenHandler.class);
+    private final DreamScreenDatagramServer server;
+    private final @Nullable DreamScreenDynamicStateDescriptionProvider descriptionProvider;
+
+    private @Nullable DreamScreenConfiguration config;
+    private int group;
+    private boolean powerOn;
+    private DreamScreenEnumMode powerOnMode = DreamScreenEnumMode.VIDEO; // TODO: Can this be persisted somewhere?
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:56]: You make it a property/parameter and update the property/parameter when it changes.
Is it also something that needs to be set every time the user changes it, or does it make more sense for the user to configure a fixed power on mode? In which case it can be a parameter of the thing.

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+			<options>
+				<option value="0">Video</option>
+				<option value="1">Music</option>
+				<option value="2">Ambient</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="scene">
+		<item-type>Number</item-type>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml:50]: Channels with options should be String.
```suggestion
		<item-type>String</item-type>
```
The numbers seem unrelated to the device itself but used as index. Maybe use the name of the enum (either in lower case) as value and then use `valueOf(` method to get the specific enum. It makes it also a bit more logical for the user.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumScene.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    static DreamScreenEnumScene fromDevice(byte deviceAmbientSceneType, byte deviceAmbientScene) {
+        return DreamScreenEnumScene.values()[deviceAmbientSceneType == 0 ? 0 : deviceAmbientScene + 1];
+    }
+
+    static DreamScreenEnumScene fromState(DecimalType state) {
+        return DreamScreenEnumScene.values()[state.intValue()];
+    }
+
+    public DecimalType state() {
+        return new DecimalType(this.ordinal());
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumScene.java:51]: This suggest the order of the enum if important, and might be tricky if someone changes the order. See my other comment in the things definition file about this.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumMode.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    public static DreamScreenEnumMode fromDevice(byte value) {
+        return value > 0 ? DreamScreenEnumMode.values()[value - 1] : null;
+    }
+
+    public static DreamScreenEnumMode fromState(DecimalType command) {
+        return DreamScreenEnumMode.values()[command.intValue()];
+    }
+
+    public DecimalType state() {
+        return new DecimalType(this.ordinal());
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumMode.java:42]: Same comment here about order.

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDynamicStateDescriptionProvider.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Dynamic channel state description provider.
+ * Overrides the state description for the controls at runtime.
+ *
+ * @author Bruce Brouwer - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = { DynamicStateDescriptionProvider.class, DreamScreenDynamicStateDescriptionProvider.class })
+public class DreamScreenDynamicStateDescriptionProvider implements DynamicStateDescriptionProvider {
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDynamicStateDescriptionProvider.java:36]: You can extends the base class implementation which handles 
```suggestion
public class DreamScreenDynamicStateDescriptionProvider extends BaseDynamicStateDescriptionProvider {
```

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+	</thing-type>
+
+	<channel-type id="power">
+		<item-type>Switch</item-type>
+		<label>Power</label>
+		<description>Turns the power to the DreamScreen on/off</description>
+	</channel-type>
+
+	<channel-type id="mode">
+		<item-type>Number</item-type>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml:37]:
```suggestion
		<item-type>String</item-type>
```

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+				<option value="6">July 4th</option>
+				<option value="7">Holiday</option>
+				<option value="8">Pop</option>
+				<option value="9">Enchanted Forest</option>
+			</options>
+		</state>
+	</channel-type>
+
+	<channel-type id="input">
+		<item-type>Number</item-type>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml:70]:
```suggestion
		<item-type>String</item-type>
```

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Here is my review.

**Comment by markus7017** [comment]
@Hilbrand rebase is done

File: bundles/org.openhab.binding.dreamscreen/src/main/feature/feature.xml-e [hunk 21]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/feature/feature.xml-e:1]: done

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:101]: thread id changed

thread terminates when serverSocket is closed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+                logger.error("No scheduler found to refresh DreamScreen TV");
+            }
+        }
+    }
+
+    private void requestRefresh() {
+        try {
+            broadcast(0xFF, 0x30, 0x01, 0x0A, new byte[0]);
+        } catch (IOException e) {
+            logger.error("Error requesting refresh of state", e);
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:118]: all logger.error() replaced by logger.debug()

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+            socket.setReuseAddress(true);
+            this.serverSocket = socket;
+
+            final Thread server = new Thread(this::runServer, "dreamscreen-tv");
+            server.setDaemon(true);
+            server.start();
+
+            final ScheduledExecutorService scheduler = this.scheduler;
+            if (scheduler != null) {
+                this.refreshing = scheduler.scheduleAtFixedRate(this::requestRefresh, 0, 10, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:107]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java (hunk 2 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:164]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+                final int len = data.getLength();
+
+                logger.debug("DreamScreen message: {} {}-{}:{}", data.getAddress(), off, len, buf);
+
+                if (isValidStateMsg(buf, off, len) && isDreamScreen(buf, off, len)) {
+                    logger.debug("Received DreamScreen message from {}", data.getAddress());
+                    refreshDreamScreen(buf, off, len, data.getAddress());
+                }
+            } catch (IOException ioe) {
+                logger.error("Error receiving DreamScreen data", ioe);
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:155]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java (hunk 3 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDatagramServer.java:224]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/i18n/dreamscreen_xx_XX.properties (hunk 5 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/i18n/dreamscreen_xx_XX.properties:1]: done

File: bundles/pom.xml (hunk 4 above)
Comment (markus7017) [reply bundles/pom.xml:1]: do you mean removing the project tag? or just adding a line break?

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenMessageConstants.java (hunk 6 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenMessageConstants.java:23]: removed, left-over

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandlerFactory.java (hunk 7 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandlerFactory.java:85]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 9 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:144]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 10 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:73]: done

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 11 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:159]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 8 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:129]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 12 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:170]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 13 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:188]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDynamicStateDescriptionProvider.java (hunk 18 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenDynamicStateDescriptionProvider.java:36]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+        } else {
+            logger.warn("DreamScreen {} is not on-line", getName());
+        }
+    }
+
+    void refreshState(final byte[] data, int off, int len, InetAddress address) {
+        this.address = address;
+        this.group = data[off + 38];
+
+        logger.trace("{}", data.toString());
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java:188]: You don't need to use `toString`:
```suggestion
        logger.trace("{}", data);
```

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Just 1 small code comment left (Besides the readme 😉)

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 19 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml:37]: Do you mean to have textual command values for mode and then map this to the enum values with valueOf()?

File: bundles/pom.xml [hunk 26]
```diff
@@ -1,4 +1,5 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (Hilbrand) [inline bundles/pom.xml:2]: The skeleton script probably changes this line. Please undo this change so it looks as it was original:
```suggestion
<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 19 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml:37]: Yes something like `<option value="video">Video</option>`. and than you can use them with `valueOf` (I put the value in lowercase, as that would be my personal preference, but requires an additional `toUppercase` before passing to `valueOf`.

The change suggestion specific here it that channels with options must be of item-type `String` and can't be `Number`.

**Comment by d8ahazard** [comment]
Fingers crossed that this goes live soon! Looking forward to controlling my DS devices with Ghome.

**Comment by d8ahazard** [comment]
@markus7017 - I see the readme update says this only supports DS devices. I sent you an email - I think I can help implement support for the rest if we have a chat.

**Comment by fwolter** [comment]
> @markus7017 - I see the readme update says this only supports DS devices, not sidekick or connect.
> I sent you an email - I think I can help implement support for the rest if we have a chat.

As the 2.5.9 merge window comes closer and close, it might be a good idea to postpone this and file another PR when OH3 has been released.

**Comment by d8ahazard** [comment]
@fwolter When's the cutoff? I'm on fire now, will have a more feature-rich PR ready in the next day or so.

**Comment by Hilbrand** [comment]
The release is planned for 20 September 2020. However it should be submitted a number of days before to be on the safe side.

**Comment by markus7017** [comment]
@Hilbrand While working on the changes I noticed that the code is not the latest version. I think I used a wrong branch as basis. Sorry, my fault.

Therefore: Let's postpone the PR here. I'll work with someone else to make it PR ready and then we could resume the PR to get it to OH 3.0.

@d8ahazard Please join us here: https://community.openhab.org/t/dreamscreen-tv-binding/53646
I could provide you a working version providing some pain relief for you :-)

**Comment by Hilbrand** [comment]
Ok. I've added the label OH3 so we can see it status.

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 19 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumMode.java (hunk 17 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumMode.java]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumScene.java (hunk 16 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenEnumScene.java]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 15 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 20 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/resources/ESH-INF/thing/thing-types.xml]: changed

File: bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java (hunk 25 above)
Comment (markus7017) [reply bundles/org.openhab.binding.dreamscreen/src/main/java/org/openhab/binding/dreamscreen/internal/DreamScreenHandler.java]: changed

File: bundles/pom.xml (hunk 26 above)
Comment (markus7017) [reply bundles/pom.xml:2]: changed

**Comment by kaikreuzer** [comment]
Closing this PR as it is not in a mergeable state. Please follow [the instructions to port this PR to the main branch](https://github.com/openhab/openhab-addons/issues/8512) for openHAB 3.
